### PR TITLE
Automatically order new pinned profile content as last (#106)

### PR DESCRIPTION
**Pull request by christophehenry**

### CODE REVIEW

File: socialhome/content/models.py [hunk 1]
```diff
@@ -118,6 +118,8 @@ def save(self, author=None, *args, **kwargs):
             # New with author, set a GUID and author
             self.guid = uuid4()
             self.author = author
+            if self.pinned:
```
Comment (jaywink) [inline socialhome/content/models.py]: Hmm this whole `elif author:` is kinda weird, had to dig to find out why it isn't just an `else:`. Anyway, your code should run even if no author. So maybe restructure like:

    if self.pk:
        ...
    else:
        if self.pinned:
            ...
        if author:
            ....

File: socialhome/content/models.py [hunk 2]
```diff
@@ -118,6 +118,8 @@ def save(self, author=None, *args, **kwargs):
...
+                self.order = Content.objects.filter(pinned=True).count() + 1
```
Comment (jaywink) [inline socialhome/content/models.py]: Your query is wrong. We don't want the count of pinned content, we want the next unused value of `order`. So you need the maximum current value and then `+1`.

**Review by jaywink — CHANGES_REQUESTED** [review]

File: socialhome/content/models.py (hunk 2 above)
Comment (christophehenry) [reply socialhome/content/models.py]: So `Content.objects.count() + 1`?

Comment (jaywink) [reply socialhome/content/models.py]: No, that would be the count of objects. You need the largest value. Check https://docs.djangoproject.com/en/1.10/ref/models/querysets/#max

File: socialhome/users/tests/test_views.py [hunk 3]
```diff
@@ -146,6 +148,18 @@ def test_contents_queryset_returns_content_in_correct_order(self, admin_client,
         assert qs[1].id == contents[1].id
         assert qs[2].id == contents[0].id
 
+    def test_content_saved_in_correct_order(self, admin_client, rf):
```
Comment (jaywink) [inline socialhome/users/tests/test_views.py]: The feature is in the model, so the test should be in content app `test_models`.

File: socialhome/users/tests/test_views.py [hunk 4]
```diff
@@ -146,6 +148,18 @@ def test_contents_queryset_returns_content_in_correct_order(self, admin_client,
...
+        request, view, contents, profile = self._get_request_view_and_content(rf, create_content=False)
+        
+        pinned_content_1 = Content(pinned=True, text="foobar")
+        pinned_content_2 = Content(pinned=True, text="foobar")
+        pinned_content_3 = Content(pinned=True, text="foobar")
+        pinned_content_1.save(author=profile)
+        pinned_content_2.save(author=profile)
+        pinned_content_3.save(author=profile)
+
+        assert {pinned_content_1.order, pinned_content_2.order, pinned_content_3.order} == {1, 2, 3}
```
Comment (jaywink) [inline socialhome/users/tests/test_views.py]: A Python set is not ordered, so this guarantees only that the elements are in the set: https://docs.python.org/3/library/stdtypes.html#set-types-set-frozenset

**Review by jaywink — CHANGES_REQUESTED** [review]

**Comment by christophehenry** [comment]
The problem is that it now overrides the `order` parameter set at creation. Like [here](https://github.com/jaywink/socialhome/blob/master/socialhome/users/tests/test_views.py#L79). I don't know how to solve.

File: socialhome/content/models.py [hunk 5]
```diff
@@ -114,10 +115,17 @@ def save(self, author=None, *args, **kwargs):
...
-            self.guid = uuid4()
-            self.author = author
+        else:
+            if author:
+                # New with author, set a GUID and author
+                self.guid = uuid4()
+                self.author = author
+
+            if self.pinned:
+                max_order = Content.objects.values_list("order").distinct().aggregate(order=Max("order"))["order"]
```
Comment (jaywink) [inline socialhome/content/models.py]: Way shorter and a lot more performant is:

    Content.objects.aggregate(Max("order"))["order__max"]

But, you're taking *all* content here. We only care about the content of current user. And only pinned content.

Ps pro-tip: use the `python manage.py shell_plus` command to locally play around with these kind of things before writing code.

File: socialhome/users/tests/test_models.py [hunk 6]
```diff
@@ -102,3 +103,14 @@ def test_get_last_name(self):
         assert profile.get_last_name() == ""
         profile.name = ""
         assert profile.get_last_name() == ""
+
+    def test_content_saved_in_correct_order(self, admin_client, rf):
```
Comment (jaywink) [inline socialhome/users/tests/test_models.py]: This is in `socialhome/users`. It should be in `socialhome/content` app `TestContentModel`.

File: socialhome/users/tests/test_views.py [hunk 7]
```diff
@@ -2,6 +2,7 @@
 from django.contrib.auth.models import AnonymousUser
 from django.core.urlresolvers import reverse
 from django.test import RequestFactory
+from factory import fuzzy
```
Comment (jaywink) [inline socialhome/users/tests/test_views.py]: Unused import.

**Review by jaywink — CHANGES_REQUESTED** [review]

**Comment by jaywink** [comment]
> The problem is that it now overrides the order parameter set at creation. Like here. I don't know how to solve.

Right, good point. The old tests need to be fixed. Easiest is to after creating the content, update it with the correct order we want, like this:

    content = ContentFactory(author=profile, pinned=True)
    Content.objects.filter(id=content.id).update(order=1)

File: socialhome/content/models.py (hunk 5 above)
Comment (christophehenry) [reply socialhome/content/models.py]: > Way shorter and a lot more performant is:
> ```
> Content.objects.aggregate(Max("order"))["order__max"]
> ```

I believe you but I don't understand. I remember during my studies my teachers telling me that filtering content like using `DISTINCT` improved queries performance.

File: socialhome/users/tests/test_models.py [hunk 8]
```diff
@@ -3,6 +3,7 @@
 
 from socialhome.users.tests.factories import ProfileFactory, UserFactory
 from socialhome.users.utils import get_pony_urls
+from socialhome.content.models import Content
```
Comment (jaywink) [inline socialhome/users/tests/test_models.py]: Unused import.

File: socialhome/content/models.py [hunk 9]
```diff
@@ -114,10 +115,17 @@ def save(self, author=None, *args, **kwargs):
...
-            self.guid = uuid4()
-            self.author = author
+        else:
+            if author:
+                # New with author, set a GUID and author
+                self.guid = uuid4()
+                self.author = author
+
+            if self.pinned:
+                max_order = Content.objects.aggregate(Max("order"))["order__max"]
```
Comment (jaywink) [inline socialhome/content/models.py]: OK this is closer, but it still needs to get content only for *user who created this current content*, not all content in the database. You need to add a filter before the aggregate to limit to current authoring user.

**Review by jaywink — CHANGES_REQUESTED** [review]
One last

**Comment by christophehenry** [comment]
Done.

**Review by jaywink — APPROVED** [review]
Great, thanks!

**Pull request merged**